src/goto-symex/features.cpp

```diff
@@ -0,0 +1,96 @@
+#include "irep2/irep2_type.h"
+#include <goto-symex/features.h>
+
+bool ssa_features::run(symex_target_equationt::SSA_stepst &steps)
+{
+
+  features.clear();
+
+  for (const auto &step : steps)
+  {
+    const expr2tc expressions[2] = {step.cond, step.guard};
+    for (const expr2tc &e : expressions)
+      check(e);
+  }
+
+  print_result();
+
+  return false;
+}
+
+void ssa_features::check(const expr2tc &e)
+{
+  if (!e)
+    return;
+
+  // TODO: We could add a cache
+  if (is_array_type(e->type))
+    features.insert(SSA_FEATURES::ARRAY);
+    
+
+  if (is_struct_type(e->type))
+    features.insert(SSA_FEATURES::STRUCTS);
+
+
+  switch (e->expr_id)
+  {
+  case expr2t::constant_fixedbv_id:
+  case expr2t::constant_floatbv_id:
+    features.insert(SSA_FEATURES::NON_INTEGER_NUMERAL);
+    break;
+    
+  case expr2t::expr_ids::mul_id:
+  case expr2t::expr_ids::div_id:
+  case expr2t::modulus_id:
+    if (!(is_constant_expr(dynamic_cast<const arith_2ops &>(*e).side_1) ||
+          is_constant_expr(dynamic_cast<const arith_2ops &>(*e).side_2)))
+      features.insert(SSA_FEATURES::NON_LINEAR);
```

Comment:
Questionable choice: Is `(unsigned short)23 * 42` (i.e. a typecast) really a non-linear expression?

Reply:
Could you clarify? You commented on modulus. Also, shouldn't be simplified by the constant folding?

Reply:
Sorry, I was referring to `mul_id`, which is also handled by this switch case, Github just defaults to 4 lines of context. Constant folding should do it, yes. Is it a necessary preprocessing step to this algorithm?

Reply:
Ow I see, you are right. Well, I could add a `contains_symbolic(const expr2tc &e)` into irep utils and use it then. Guess that is safer.



> Sorry, I was referring to `mul_id`, which is also handled by this switch case, Github just defaults to 4 lines of context. Constant folding should do it, yes. Is it a necessary preprocessing step to this algorithm?

Ow I see, you are right. Well, I could add a `contains_symbolic(const expr2tc &e)` into irep utils and use it then. Guess that is safer.

Reply:
Well, the question is, what "SSA contains non-linear features" really means. Consider for instance `a := b / 0`. This cannot be simplified because it's undefined. But if `b` was a constant, contains_symbolic() would return false. How would that correspond to "contains non-linear features"? Maybe an easier approach is to call the simplification directly and ask whether the result is "entirely constant" meaning, that it positively only contains constant_*2t expressions and defined functions (such as typecasts and no signed overflows). This predicate could be extended later, no need to get it perfect right now in this PR.